File: client/tests/utils/quiz-test-runner.js
import { startQuiz } from './quiz-runners/startQuiz';
import { completePersonalitySection } from './quiz-runners/personalitySection';
import { completeSubjectSection } from './quiz-runners/subjectSection';
import { 
  handlePersonalityTieBreaker 
} from './quiz-runners/personalityTieBreaker';
import { 
  handleSubjectTieBreaker 
} from './quiz-runners/subjectTieBreaker';
import { verifyResultsText } from './quiz-runners/resultsTextCheck';
import { verifyPieChartToggle } from './quiz-runners/verifyPieChartToggle';
import { verifyCareerRecommendations } from './quiz-runners/careerRecommendation';
import { verifyNetworkBoard } from './quiz-runners/networkBoard';
import { deleteUserResult } from './quiz-runners/deleteUserResult';
import { checkContextRestarted } from './quiz-runners/checkContextRestarted';
import { checkQuizRestart } from './quiz-runners/checkQuizRestart';

export async function runQuizTestCase(page, testCase, subjectsData) {
  await startQuiz(page);
  await completePersonalitySection(page, testCase.personalityAnswers);
  
  const tieBreakerVisible = await page.isVisible('h2:has-text("We Found a Tie!")');
  if (tieBreakerVisible) {
    if (!testCase.preferredTrait) {
      throw new Error('Tie breaker screen shown but no preferred trait specified in test case');
    }
    await handlePersonalityTieBreaker(page, testCase.preferredTrait);
  } else {
    await page.waitForSelector('h2:has-text("STEAM Subject Quiz")', { 
      timeout: 2000,
      state: 'visible'
    });
  }
  
  await completeSubjectSection(page, testCase.subjectAnswers, subjectsData);
  
  const subjectTieBreakerVisible = await page.isVisible('h2:has-text("Your Strongest Subjects")');
  if (subjectTieBreakerVisible) {
    if (!testCase.preferredSubject) {
      throw new Error('Subject tie breaker screen shown but no preferred subject specified in test case');
    }
    await handleSubjectTieBreaker(page, testCase.preferredSubject);
  }

  await verifyResultsText(page, testCase);
  
  // Verify pie chart toggle functionality
  try {
    await verifyPieChartToggle(page);
  } catch (error) {
    console.error('Pie chart toggle verification failed:', error);
    throw error;
  }

  await verifyCareerRecommendations(page, testCase);
  try {
    await verifyNetworkBoard(page, testCase);
  } catch (error) {
    console.error('Network Board verification failed:', error);
    throw error;
  }

  await deleteUserResult(page, testCase);
  await checkQuizRestart(page);
  await checkContextRestarted(page, testCase);
}

File: client-android/app/src/androidTest/java/com/quiz/utils/QuizTestRunner.kt
/**
 * Ports from: client/tests/utils/quiz-test-runner.js
 * Reviewed as ported correctly 10/10/2024
 */
object QuizTestRunner {
    suspend fun runQuizTestCase(
        composeRule: ComposeTestRule,
        testCase: QuizTestCases.TestCase,
        subjectsData: Map<QuizTestCases.Subject, SubjectSectionRunner.SubjectData>,
        feedbackDatabase: Map<String, CareerRecommendationRunner.CareerFeedback>,
        quizContext: ContextRestartRunner.QuizContext
    ) {
        with(composeRule) {
            // Start quiz
            StartQuizRunner.startQuiz(this, testCase.username)

            // Complete personality section
            PersonalitySectionRunner.completePersonalitySection(this, testCase.personalityAnswers)

            // Handle personality tie breaker if needed
            if (hasPersonalityTie()) {
                require(testCase.preferredTrait != null) {
                    "Tie breaker screen shown but no preferred trait specified in test case"
                }
                PersonalityTieBreakerRunner.handleTie(this, testCase.preferredTrait)
            } else {
                // Wait for subject quiz section if no tie
                onNodeWithText("STEAM Subject Quiz")
                    .assertIsDisplayed()
            }

            // Complete subject section
            SubjectSectionRunner.completeSubjectSection(this, testCase.subjectAnswers, subjectsData)

            // Handle subject tie breaker if needed
            if (hasSubjectTie()) {
                require(testCase.preferredSubject != null) {
                    "Subject tie breaker screen shown but no preferred subject specified in test case"
                }
                SubjectTieBreakerRunner.handleTie(this, testCase.preferredSubject)
            }

            // Verify results
            ResultsTextCheckRunner.verifyResultsText(this, testCase)
            CareerRecommendationRunner.verifyCareerRecommendations(this, testCase, feedbackDatabase)

            try {
                NetworkBoardRunner.verifyNetworkBoard(this, testCase)
            } catch (e: AssertionError) {
                Log.e("QuizTestRunner", "Network Board verification failed: ${e.message}")
                throw e
            }

            // Cleanup
            DeleteUserResultRunner.deleteUserResult(this, testCase)
            QuizRestartRunner.verifyQuizRestart(this)
            ContextRestartRunner.verifyContextRestarted(this, quizContext)
        }
    }

    private fun ComposeTestRule.hasPersonalityTie(): Boolean {
        return try {
            onNodeWithText("We Found a Tie!")
                .assertExists()
            true
        } catch (e: AssertionError) {
            false
        }
    }

    private fun ComposeTestRule.hasSubjectTie(): Boolean {
        return try {
            onNodeWithText("Your Strongest Subjects")
                .assertExists()
            true
        } catch (e: AssertionError) {
            false
        }
    }
} 